### PR TITLE
[Function] 1. Refactor the code of function overload by SFINE 2. Fix the bug of hex function

### PR DESCRIPTION
**Pull request by HappenLee**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by zbtzbtzbt — COMMENTED** [review]
[here1](https://github.com/doris-vectorized/doris-vectorized/blob/0ba567c7256243d278685c716e8d35d872f456dd/be/src/vec/data_types/data_type.h#L44-L46)、[here2](https://github.com/doris-vectorized/doris-vectorized/blob/0ba567c7256243d278685c716e8d35d872f456dd/be/src/vec/data_types/data_type.cpp#L36) 这两处一起删了吧

**Comment by zbtzbtzbt** [comment]
[here3](https://github.com/doris-vectorized/doris-vectorized/blob/0ba567c7256243d278685c716e8d35d872f456dd/be/src/vec/functions/function.h#L279) 还有这一行，也删了吧

File: be/src/vec/functions/function.h [hunk 1]
```diff
@@ -29,6 +29,11 @@ namespace doris::vectorized {
 
 class Field;
 
+// Only use dispose the variadic argument
+template <typename T>
+auto has_variadic_argument_types(T&& arg) -> decltype(T::get_variadic_argument_types()) {};
```
Comment (zbtzbtzbt) [inline be/src/vec/functions/function.h:34]: 请问这里为啥用右值引用呢

Comment (HappenLee) [reply be/src/vec/functions/function.h:34]: 因为std::declval返回右值

File: be/src/vec/functions/function_string.cpp [hunk 2]
```diff
@@ -266,6 +266,53 @@ struct ReverseImpl {
...
+    static constexpr auto name = "hex";
+};
+
+struct HexStringImpl {
+    static DataTypes get_variadic_argument_types() {
+        return {std::make_shared<vectorized::DataTypeString>()};
+    }
+
+    static size_t hex_encode(const char* src_str, size_t length, char* dst_str) {
+        static const char* hex_table = "0123456789abcdef";
```
Comment (zbtzbtzbt) [inline be/src/vec/functions/function_string.cpp:279]: 这里用大写字母吧

File: be/src/vec/functions/math.cpp [hunk 3]
```diff
@@ -171,47 +171,46 @@ struct HexIntImpl {
...
     
     static Status vector(const ColumnInt64::Container& data, ColumnString::Chars& res_data,
                          ColumnString::Offsets& res_offsets) {
-        
         res_offsets.resize(data.size());
         size_t input_size = res_offsets.size();
+        char ans[17];
         for (size_t i = 0; i < input_size; ++i) {
-            StringOP::push_value_string(hex(data[i]), i, res_data, res_offsets);
+            StringOP::push_value_string(hex(data[i], ans), i, res_data, res_offsets);
```
Comment (zbtzbtzbt) [inline be/src/vec/functions/math.cpp:207]: 有个warning
```[363/366] Building CXX object src/vec/CMakeFiles/Vec.dir/functions/math.cpp.o
../src/vec/functions/math.cpp: In static member function 'static std::string_view doris::vectorized::HexIntImpl::hex(uint64_t, char*)':
../src/vec/functions/math.cpp:180:44: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
  180 |         static constexpr char* hex_table = "0123456789ABCDEF";
      |                                            ^~~~~~~~~~~~~~~~~~
../src/vec/functions/math.cpp:198:22: warning: narrowing conversion of 'i' from 'int' to 'std::basic_string_view<char>::size_type' {aka 'long unsigned int'} [-Wnarrowing]
  198 |         return {ans, i};```

File: be/src/vec/functions/math.cpp [hunk 4]
```diff
@@ -171,47 +171,46 @@ struct HexIntImpl {
...
-        char ans[17];
+        if (num == 0) { return {hex_table, 1};}
+
         int i = 0;
         while (num) {
-            ans[i++] = hexTable[num & 15];
+            ans[i++] = hex_table[num & 15];
             num = num >> 4;
         }
         ans[i] = '\0';
```
Comment (zbtzbtzbt) [inline be/src/vec/functions/math.cpp:189]: 这里用string_view的话，这行应该没用

Comment (HappenLee) [reply be/src/vec/functions/math.cpp:189]: 这个先保持这样吧

**Pull request merged**